### PR TITLE
Spot module as a separate package

### PR DESCRIPTION
**Pull request by anjannath**
depends on #277

### CODE REVIEW

File: pkg/provider/util/instancetypes/instancetypes.go [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+}
+
+func validate(cpus, memory int32, arch arch) error {
+	if cpus > 0 && memory > 0 && arch.String() != "" {
+		return nil
+	}
+	return fmt.Errorf("Invalid values for CPUs: %d, Memory: %d and Arch: %s", cpus, memory, arch)
+}
+
+func (r *AwsInstanceRequest) GetMachineTypes() ([]string, error) {
```
Comment (adrianriobo) [inline pkg/provider/util/instancetypes/instancetypes.go:53]: Do you need to rebase and adapt for this?

**Comment by adrianriobo** [comment]
Also same comment as per type of machine applies here, if we rely on this module; is it only SDKs dependent?

File: pkg/provider/util/instancetypes/instancetypes.go (hunk 1 above)
Comment (anjannath) [reply pkg/provider/util/instancetypes/instancetypes.go]: it is rebased and ready for review now :)

**Comment by anjannath** [comment]
> Also same comment as per type of machine applies here, if we rely on this module; is it only SDKs dependent?

yes, this also only depends on the sdks, i've added example on how to use it in: b553996b7071adfcc624d627320557eaa49e97ea (not sure we should merge this or not, i am okay to drop it from the PR if needed)

File: examples/spotprice/main.go [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+		MemoryGib:  8,
+		Arch:       "amd64",
+		NestedVirt: false,
+		Os:         "fedora",
+		OSVersion:  "39",
+	}
+
+	// Get the lowest price for only aws
+	//spi, err := spotReq.GetAwsLowestPrice()
+
```
Comment (adrianriobo) [inline examples/spotprice/main.go:34]: I guess this is expecting to have the specific Envs with credentials ...can we add that here? I mean os.SetEnv(AWS..., SampleSecretKey)....

To make more clear how we are expecting the credentials to be set?

Comment (anjannath) [reply examples/spotprice/main.go:47]: updated with the required env variables that needs to be set for both aws and azure

**Comment by anjannath** [comment]
> Also I just merged the defaults so probably this will need rebase again ?

thanks, rebased on main

**Pull request merged**